### PR TITLE
Base container

### PR DESCRIPTION
**Pull request by drivera-armedia**
First version of the container to be merged to the main branch

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1 +1,41 @@
+ARG ARCH="amd64"
+ARG OS="linux"
+FROM 345280441424.dkr.ecr.ap-south-1.amazonaws.com/ark_base:latest
+LABEL maintainer="Armedia, LLC"
```
Comment (fabricetriboix) [inline Dockerfile]: We have a couple of other labels, example for `ark_activemq`:

```
LABEL   ORG="Armedia LLC" \
        APP="ActiveMQ" \
        VERSION="1.1" \
        IMAGE_SOURCE="https://github.com/ArkCase/ark_activemq" \
        MAINTAINER="Armedia LLC"
```

File: Dockerfile [hunk 2]
```diff
@@ -1 +1,41 @@
...
 
+ARG ARCH="amd64"
```
Comment (fabricetriboix) [inline Dockerfile:6]: This line is a duplicate of line 1

File: Dockerfile [hunk 3]
```diff
@@ -1 +1,41 @@
...
+ARG OS="linux"
```
Comment (fabricetriboix) [inline Dockerfile:7]: Duplicate of line 2

File: Dockerfile [hunk 4]
```diff
@@ -1 +1,41 @@
...
+ARG VER="2.28.1"
+ARG PKG="prometheus"
+ARG SRC="${PKG}-${VER}.${OS}-${ARCH}"
+
+RUN curl \
+		-L "https://github.com/prometheus/${PKG}/releases/download/v${VER}/${SRC}.tar.gz" \
+		-o package.tar.gz && \
```
Comment (fabricetriboix) [inline Dockerfile]: We decided to store the artifacts in our own S3 bucket: `arkcase-container-artifacts`
We also have a `get-artifacts.sh` script in each repo to download the artifacts from the S3 bucket prior to building the Docker image (this is because otherwise we would need to have the awscli installed inside the container during the Docker build, along with some credentials and that was seen as too complicated).

File: Dockerfile [hunk 5]
```diff
@@ -1 +1,41 @@
...
+	tar -xzvf "package.tar.gz" && \
+	mkdir -pv "/prometheus" && \
+	mkdir -pv "/etc/prometheus" && \
+	mkdir -pv "/usr/share/prometheus" && \
```
Comment (fabricetriboix) [inline Dockerfile]: I am personally a fan of keeping everything related to the containerized app in a single directory, such as `/app`. But this is just a comment from me, I don't think we have a policy regarding this.

File: Dockerfile [hunk 6]
```diff
@@ -1 +1,41 @@
...
+	mv -vif "${SRC}/LICENSE"                     "/LICENSE" && \
+	mv -vif "${SRC}/NOTICE"                      "/NOTICE" && \
+	mv -vif "${SRC}/prometheus"                  "/bin/prometheus" && \
+	mv -vif "${SRC}/promtool"                    "/bin/promtool" && \
+	mv -vif "${SRC}/prometheus.yml"              "/etc/prometheus/prometheus.yml" && \
+	mv -vif "${SRC}/console_libraries/"          "/usr/share/prometheus/" && \
+	mv -vif "${SRC}/consoles/"                   "/usr/share/prometheus/" && \
+	ln -s "/usr/share/prometheus/console_libraries" "/etc/prometheus" && \
+	ln -s "/usr/share/prometheus/consoles"          "/etc/prometheus" && \
+	chown -R nobody:nobody "/etc/prometheus" "/prometheus" && \
```
Comment (fabricetriboix) [inline Dockerfile]: I would prefer to see `root:root` here, as it is more restrictive. But we had some discussions about this and it seems that Armedia is OK to be not very tight on file/directory permissions. So just take this as a comment/suggestion.

File: Dockerfile [hunk 7]
```diff
@@ -1 +1,41 @@
...
+	rm -rvf "${SRC}" package.tar.gz
+
+#COPY npm_licenses.tar.bz2                        /npm_licenses.tar.bz2
+
+USER       nobody
```
Comment (fabricetriboix) [inline Dockerfile]: Please create a user using `useradd` rather than using `nobody` (which is now promoted to somebody)

File: Dockerfile [hunk 8]
```diff
@@ -1 +1,41 @@
...
+EXPOSE     9090
+VOLUME     [ "/prometheus" ]
```
Comment (fabricetriboix) [inline Dockerfile]: This is the first time I see this `VOLUME` command in a Dockerfile. Could you please explain to me what it does? Thanks

File: Dockerfile [hunk 9]
```diff
@@ -1 +1,41 @@
...
+WORKDIR    /prometheus
+ENTRYPOINT [ "/bin/prometheus" ]
```
Comment (fabricetriboix) [inline Dockerfile]: Please do not override the `ENTRYPOINT`. There must be a very good reason to do that and it usually is a bad idea in 99% of the cases.

**Review by fabricetriboix — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 6 above)
Comment (drivera-armedia) [reply Dockerfile]: The value of using an exact user/group here is that then you can grant extra-container access to container files (in mounted volumes, for instance) without requiring root access.

File: Dockerfile (hunk 7 above)
Comment (drivera-armedia) [reply Dockerfile]: The user "nobody" is part of the base CentOS 7 install, and is meant precisely for this sort of thing. Are you saying we should actually have a specific "prometheus" user (in this case)?

File: Dockerfile (hunk 8 above)
Comment (drivera-armedia) [reply Dockerfile]: The VOLUME command adds persistence beyond the container. I.e. the container's FS is created, and if a volume isn't attached to it as part of the container's startup, then one is created. This volume will contain persistent files separate from the container's so if the container goes down, the volume remains and can be re-used by a later container if desired.

The simpler way to use this is by defining named volumes which can easily be re-used (b/c their names are easily recognizable). This is how you can achieve data persistence for database containers, etc. Alfresco does this for its data store, for instance...

File: Dockerfile (hunk 9 above)
Comment (drivera-armedia) [reply Dockerfile]: I'm confused ... nearly every Dockerfile I've seen uses ENTRYPOINT vs. CMD ... are you saying that you prefer the use of CMD to ENTRYPOINT?

From the docs: "The main purpose of a CMD is to provide defaults for an executing container. These defaults can include an executable, or they can omit the executable, in which case you must specify an ENTRYPOINT instruction as well."

This suggests that CMD isn't the preferred way to specify a container's primary executable, but rather a possibility to do so if one chooses to not use ENTRYPOINT as well ...

File: Dockerfile (hunk 4 above)
Comment (drivera-armedia) [reply Dockerfile]: I see the value in this, but the intent here is to get it from the source - i.e. "unadulterated", if you will...

File: Dockerfile (hunk 1 above)
Comment (drivera-armedia) [reply Dockerfile]: Added

File: Dockerfile (hunk 2 above)
Comment (drivera-armedia) [reply Dockerfile:6]: Consolidated

File: Dockerfile (hunk 3 above)
Comment (drivera-armedia) [reply Dockerfile:7]: Consolidated

File: Dockerfile (hunk 4 above)
Comment (fabricetriboix) [reply Dockerfile]: What Jon suggested is that we should be able to build our images even if the original sources are not available (server down or other things). Also, if we pull a lot of images from Docker Hub (esp. from the CI/CD pipelines), we might hit Docker Hub rate limits.

So what we do is download the "unadulterated" package and upload it to the S3 bucket. Then the `get-artifacts.sh` script downloads the artifacts from the S3 bucket.

File: Dockerfile (hunk 6 above)
Comment (fabricetriboix) [reply Dockerfile]: @drivera-armedia I don't understand. Could you please expand or provide some examples?

File: Dockerfile (hunk 7 above)
Comment (fabricetriboix) [reply Dockerfile]: My understanding of Linux is that the user `nobody` is used to isolate things. Here Prometheus will run as `nobody` and thus will have read/write access to `/etc/prometheus` and `/prometheus`, which is bad. Prometheus should have read access only to what it needs to read, and write access only to what it needs to write (which I suppose is going to be only its database). 

Regarding the name of the user, I think that yes, `prometheus` would be good and in line with what we did so far. But if you want to run Prometheus as `nobody`, that's fine to me as long as `nobody` is not the owner for `/etc/prometheus`, `/prometheus`, and anything else that the Prometheus process does not need to write to.

File: Dockerfile (hunk 8 above)
Comment (fabricetriboix) [reply Dockerfile]: All right, I see. So if Kubernetes already mounts a volume on `/prometheus`, Docker will not attempt to create one if its own volume, is that correct?

Comment (drivera-armedia) [reply Dockerfile]: Correct. You use "VOLUME" to indicate that the data contained within is to be persisted independently of the container's own persistence. I.e. you nuke the container, the volume stays so you can re-use it or recover its contents later. You can still do this by hard-mounting volumes unto directories, but it's generally a good idea to declare that your container expects a (set of) volume(s) to be provisioned for persistent storage.

File: Dockerfile (hunk 7 above)
Comment (drivera-armedia) [reply Dockerfile]: Well, /etc/prometheus and /prometheus (being changed to /app/conf and /app/data respectively) are the only areas the process should read and write from explicitly.

I was seeking to avoid having to create users and start "littering" the exported volumes with UIDs that won't exist outside. Furthermore, if each container creates its own user there's a chance all the UIDs will be the same (b/c they're the first user created in the OS image), so that kind of defeats the purpose of isolation in the case of volumes...

File: Dockerfile (hunk 9 above)
Comment (fabricetriboix) [reply Dockerfile]: `CMD` is totally the preferred way to specify what you want to run inside your Docker container. I my opinion, there are only 2 valid reasons why you would want to override `ENTRYPOINT`:

1. You want to provide the user with a convenient way to use the same Docker image for two different purposes. For example, I think the official MariaDB image can run either as client or server, and they override `ENTRYPOINT` to achieve this.
2. You absolutely need to run some script when your Docker image is started, even if you run `docker run -it MYIMAGE sh`.

By default, `ENTRYPOINT` is set to `/bin/sh -c`, which allows to run anything. If you set it to, say, `/bin/prometheus`, then you won't be able to run, say, `docker run -it MYIMAGE sh`, because `sh` is not an argument for `/bin/prometheus`. You can still manage to run `docker run` by overriding the overriden `ENTRYPOINT` and set it back to `/bin/sh -c`, but it's just as simple to not override `ENTRYPOINT` in the first place.

In any case, I don't see any advantage of overriding `ENTRYPOINT` in 99% of cases.

File: Dockerfile (hunk 6 above)
Comment (drivera-armedia) [reply Dockerfile]: This is about mounted volumes. If I mount a directory unto a container, its contents will have to have permissions such that a) the running process can read/write as appropriate to those locations, and b) a user outside in the host machine can read/write as appropriate to those locations.

If those directories are owned by root (UID 0), then the process inside the container must run as root. If the process were compromised, then that process could do a LOT of damage within the container. While the container ... _contains_ the damage, you have a 2nd problem: to access the data stored within those mounted volumes, you'd have to be root yourself which tends to be a very bad thing to spread around.

So, by using a lowered user we can maintain isolation while also gaining access as needed, but without needing to be root or resorting to granting global read/write/execute permissions.

File: Dockerfile (hunk 9 above)
Comment (drivera-armedia) [reply Dockerfile]: Fair enough. I'll adjust as needed. I don't think I've seen this issue, though - I'm fully able to launch shells on images which have an ENTRYPOINT declared.

As I read the docs, ENTRYPOINT is meant to point at the executable which is to be run by the container, and CMD is meant to give the default parameters to that executable.

File: Dockerfile (hunk 5 above)
Comment (JonHolman) [reply Dockerfile]: we are doing that as a standard, so it should be a policy

File: Dockerfile [hunk 10]
```diff
@@ -1 +1,47 @@
...
+	chown -R nobody:nobody	"/app/data" "/app/conf" && \
+	chmod -R ug+rwX,o-rwx	"/app/data" "/app/conf" && \
+	rm -rvf "${SRC}" "package.tar.gz"
+
+#COPY npm_licenses.tar.bz2	/npm_licenses.tar.bz2
+
+USER		nobody
+EXPOSE		9090
+VOLUME		[ "/app/data", "/app/conf" ]
+WORKDIR		/app/data
```
Comment (JonHolman) [inline Dockerfile]: as @fabricetriboix said on the other PR, is volume supposed to be workdir as well?

File: Dockerfile [hunk 11]
```diff
@@ -1 +1,47 @@
...
+	mv -vif "${SRC}/consoles/"						"/usr/share/prometheus/" && \
+	ln -s "/usr/share/prometheus/console_libraries"	"/app/conf" && \
+	ln -s "/usr/share/prometheus/consoles"			"/app/conf" && \
+	chown -R nobody:nobody	"/app/data" "/app/conf" && \
+	chmod -R ug+rwX,o-rwx	"/app/data" "/app/conf" && \
+	rm -rvf "${SRC}" "package.tar.gz"
+
+#COPY npm_licenses.tar.bz2	/npm_licenses.tar.bz2
+
+USER		nobody
```
Comment (JonHolman) [inline Dockerfile]: As @fabricetriboix said on the other PR, please don't use nobody.  Create a user and use it.

File: Dockerfile [hunk 12]
```diff
@@ -1 +1,47 @@
...
+EXPOSE		9090
+VOLUME		[ "/app/data", "/app/conf" ]
+WORKDIR		/app/data
+ENTRYPOINT	[ "/bin/prometheus" ]
```
Comment (JonHolman) [inline Dockerfile]: as @fabricetriboix said on the other PR, please don't use ENTRYPOINT

File: Dockerfile [hunk 13]
```diff
@@ -1 +1,47 @@
...
+	mv -vif "${SRC}/prometheus.yml"					"/app/conf/prometheus.yml" && \
+	mv -vif "${SRC}/console_libraries/"				"/usr/share/prometheus/" && \
+	mv -vif "${SRC}/consoles/"						"/usr/share/prometheus/" && \
+	ln -s "/usr/share/prometheus/console_libraries"	"/app/conf" && \
+	ln -s "/usr/share/prometheus/consoles"			"/app/conf" && \
+	chown -R nobody:nobody	"/app/data" "/app/conf" && \
+	chmod -R ug+rwX,o-rwx	"/app/data" "/app/conf" && \
+	rm -rvf "${SRC}" "package.tar.gz"
+
+#COPY npm_licenses.tar.bz2	/npm_licenses.tar.bz2
```
Comment (JonHolman) [inline Dockerfile]: why is this here but commented out?

File: Dockerfile [hunk 14]
```diff
@@ -1 +1,47 @@
...
+	ln -sv "/app/conf" "/etc/prometheus" && \
+	mv -vif "${SRC}/LICENSE"						"/LICENSE" && \
+	mv -vif "${SRC}/NOTICE"							"/NOTICE" && \
+	mv -vif "${SRC}/prometheus"						"/bin/prometheus" && \
+	mv -vif "${SRC}/promtool"						"/bin/promtool" && \
+	mv -vif "${SRC}/prometheus.yml"					"/app/conf/prometheus.yml" && \
+	mv -vif "${SRC}/console_libraries/"				"/usr/share/prometheus/" && \
+	mv -vif "${SRC}/consoles/"						"/usr/share/prometheus/" && \
+	ln -s "/usr/share/prometheus/console_libraries"	"/app/conf" && \
+	ln -s "/usr/share/prometheus/consoles"			"/app/conf" && \
```
Comment (JonHolman) [inline Dockerfile]: "                              "

Comment (JonHolman) [reply Dockerfile]: that was supposed to have a lot of spaces between the quotes! to point out the large space between console and /app/conf

**Review by JonHolman — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 9 above)
Comment (drivera-armedia) [reply Dockerfile]: Ok so here's a concrete example of why using CMD as the base vs. ENTRYPOINT is a bad idea ... take this docker-compose.yml fragment:

```
`version: "3"

services:
    prometheus:
        image: backups.rivera.prv:6088/ark_prometheus:latest
        container_name: prometheus
        dns_search: rivera.prv
        dns:
            - 192.168.3.253
            - 192.168.3.252
        volumes:
            - ./data/prometheus:/app/data
            - ./conf/prometheus:/app/conf
        ports:
            - 9090:9090
        depends_on:
            - cadvisor
            - pushgateway
        restart: unless-stopped
        command:
            - '--config.file=/app/conf/prometheus.yml'
            - '--storage.tsdb.path=/app/data'
            - '--web.console.libraries=/usr/share/prometheus/console_libraries'
            - '--web.console.templates=/usr/share/prometheus/consoles'
```

When using CMD, the command would also have to include the executable one wishes to run. That kind of defeats the purpose of having *application containers*. With ENTRYPOINT, the above works just fine b/c it assumes the base command that will be used, and thus one can limit oneself to tuning parameters as appropriate.

This also means that when relying on CMD instead of ENTRYPOINT, a mistake on a developer's part can lead to the wrong executable being launched on an image ...

I'm going to keep ENTRYPOINT until we figure this out. Using CMD in the manner suggested is, in my mind, a BAD idea. I hadn't found a reason to oppose it, until right this second when I re-tried to rebuild the actual stack these images will be a part of.

Hopefully you guys have a solution for the above scenario and we can all come to a happy convergence :).

File: Dockerfile (hunk 10 above)
Comment (drivera-armedia) [reply Dockerfile]: Why would this be an issue? This is application-specific, and thus really shouldn't be a point of review ...

File: Dockerfile (hunk 12 above)
Comment (drivera-armedia) [reply Dockerfile]: Until a resolution is found, ENTRYPOINT will remain in use for the reasons stated above.

File: Dockerfile (hunk 13 above)
Comment (drivera-armedia) [reply Dockerfile]: B/c I'm lazy and never got around to removing it :D  (it's gone on the commit history now, btw).

File: Dockerfile (hunk 6 above)
Comment (fabricetriboix) [reply Dockerfile]: @drivera-armedia That seems a bit obscure to me... Generally speaking, I think anything that must be writeable by the containerized app should be isolated in separate directories. For example with ActiveMQ, I have `/app/conf` for the configuration files which all owned by root and read-only for the `activemq` user. Ditto for other files/directories that ActiveMQ needs to be able to access only on a read-only basis. For files/directories that ActiveMQ needs to be able to write to, that's under `/app/data`, which is owned by the `activemq` user and with permissions to write. When the container is started, Kubernetes will mount a persistent volume on `/app/data`. 

In your example, I still don't understand why, for example, `/etc/prometheus` should be owned by `nobody`. I Prometheus will not write into this directory (at least I think it should not). I guess you will have only static configuration files in there. Are saying that you want to mount a volume on `/etc/prometheus`? Even if that's the case, I don't understand why it should be owned by `nobody`...

At the end of the day, whether `/etc/prometheus` is owned by root or by any other user, that's fine with me. The only thing I am saying is that the Prometheus process should not have write access to it. Ditto for any other file/directory that Prometheus should only access on a read-only basis.

Comment (drivera-armedia) [reply Dockerfile]: The user _**nobody**_ was used out of the practicality of using an existing user without having to create one. At the O/S level there's no difference between using UID X or UID Y as long as neither is "0", and neither clashes with an existing host system user (i.e. 501).

There's no problem in creating a user for prometheus with a given UID, except the aforementioned potential for collision with an existing host O/S user which would grant access to someone not meant to gain that access.  If we define a UID range to use for creating our users, then definitely this is something to consider ... but now we would have to maintain a list of users.

Say we deploy Prometheus alongside ActiveMQ in the same machine. It just so happens that when the containers' **prometheus** and **activemq** users were created, they both got the same UID ... and it just so happens that there's a user in the host system with that UID already. This means that this outside user already - albeit likely unknowingly - has access to both Pometheus's and ActiveMQ's data volumes...

Granted, we're working in K8S here and things are different. But the concept is the same. The user "nobody" can safely be assumed to not be assigned to any user, and is generally designed to be used as an "isolation" user already. We can definitely use other users as needed, but in my mind we're solving a problem we don't really have ... unless we were to decide to keep a catalog of container users and adhere to it very strictly ...

File: Dockerfile (hunk 7 above)
Comment (fabricetriboix) [reply Dockerfile]: @drivera-armedia I am quite sure that Docker does not permit one container to see volumes that are not mounted to it, whether the UIDs are the same or not. Also, I don't think we should preoccupy ourselves with users on the host, which might be minikube, a docker-compose running on Windows, an EKS node (derived for Amazon Linux), etc.

You say that `/etc/prometheus` must be writeable by Prometheus? That sounds very odd to me... Could you please elaborate?

`/prometheus` is where Prometheus will store its database, is that correct?

I think creating a dedicated user is more explicit, and also allows you more control over this user (eg: location of the home directory, which shell to run, whether you want to disallow shell access entirely for this user, etc.)

File: Dockerfile (hunk 8 above)
Comment (fabricetriboix) [reply Dockerfile]: @drivera-armedia All right, that sounds good. I learned something new today!

File: Dockerfile (hunk 9 above)
Comment (fabricetriboix) [reply Dockerfile]: @drivera-armedia Sure, I see your point... I see pros and cons for both aspects. I am kind of changing my mind on this and I tend to think you are right.

File: Dockerfile (hunk 12 above)
Comment (fabricetriboix) [reply Dockerfile]: @JonHolman Actually, I would tend to think Diego is right on this point.

File: Dockerfile (hunk 7 above)
Comment (drivera-armedia) [reply Dockerfile]: If you're concerned about denying write-access, one can always mount volumes in read-only mode, regardless of the actual permissions granted at the O/S level. I think this is threading a bit thinly in the sense that we're solving a problem we've yet to have, and I don't think we have a pressing need to resolve now (not to mention that the aforementioned read-only volume mounting would solve it summarily with no change to permissions).

The Dockerfile has been modified to use /app/conf and /app/data, which should match what you have in other places.  The value of ensuring that the UID doesn't collide with a host O/S UID is to help minimize inadvertent modifications of the contained data: you want to change it? you'll need to make a concerted effort to change it ...

Here's an example of mounting volumes in docker-compose.yml - note the 4th volume being mounted as read-only:

image

File: Dockerfile (hunk 6 above)
Comment (fabricetriboix) [reply Dockerfile]: @drivera-armedia We should not try to anticipate where the container will run. I think we should not make any assumptions about features of the host OS that will run the container. I can setup a host OS where my user id will be the same number as the `nobody` user of the container. Securing access to volumes from the host OS is not the concern of the container. I totally fail to understand what's the point of keeping a "catalog of container users". User `nobody` or any other user are the same as far as the kernel is concerned, except for user id 0 (and I am paraphrasing what you wrote). 

What's your take on ensuring the containerized app has read-only access to files/directories it only needs to read and write access only to files/directories is has to be able to write to?

File: Dockerfile (hunk 7 above)
Comment (fabricetriboix) [reply Dockerfile]: @drivera-armedia What about denying write access to a file/directory that is not mounted?

The "problem" I am trying to solve is us adhering to best practices, one of the most important being the principle of least privilege.

File: Dockerfile (hunk 6 above)
Comment (drivera-armedia) [reply Dockerfile]: It's a very beautiful ideal, but it requires two things: a) extra work, and b) perfect certainty that this will not hobble the application in some unforeseen way (i.e. b/c it has a bug that tries to write to an anticipated read-only area).

I see the value, but not the immediate need. Or, more accurately, I see too much effort being required to make sure we can leverage this safely without risking unforeseen consequences - effort that we can better expend elsewhere.

File: Dockerfile (hunk 7 above)
Comment (drivera-armedia) [reply Dockerfile]: I get that. But as I mentioned elsewhere, unless we have perfect knowledge that the application will never attempt to write to a read-only volume, or that any such attempt will not result in a problem, I think it's a beautiful yet risky concept to pursue at this time.

Finally, you're never going to be able to deny access to a file/directory that's not mounted unless you resort to permissions trickery, which can get gnarly.  In some ways you want to treat the contents of the container as its own independent machine, while in others you obviously want to do the opposite ... you want to rely on VOLUME for persistence, and rely on the ability to quickly re-create a corrupted container from scratch when there's a problem. This means that you don't really want to worry too much about limiting who modifies what - unless, of course, this is in a persistent area, which almost certainly needs to be writable by the application anyway...

File: Dockerfile (hunk 14 above)
Comment (drivera-armedia) [reply Dockerfile]: Yeah this was an attempt at aligning the 2nd parameters ... came up with a different approach

File: Dockerfile (hunk 12 above)
Comment (JonHolman) [reply Dockerfile]: Thanks @fabricetriboix  please update our best practices confluence page.

File: Dockerfile [hunk 15]
```diff
@@ -1,73 +1,78 @@
...
-ARG UID="nobody"
-ARG GID="nobody"
+ARG UID="prometheus"
 
+#
+# Some important labels
+#
 LABEL ORG="Armedia LLC"
-LABEL MAINTAINER="Armedia LLC"
+LABEL MAINTAINER="Devops Team <devops@armedia.com>"
```
Comment (fabricetriboix) [inline Dockerfile]: Maybe "Armedia DevOps Teams"?

**Review by fabricetriboix — DISMISSED** [review]

File: Dockerfile (hunk 15 above)
Comment (drivera-armedia) [reply Dockerfile]: Sure!! I'll fix it.

File: Dockerfile (hunk 7 above)
Comment (fabricetriboix) [reply Dockerfile]: > I get that. But as I mentioned elsewhere, unless we have perfect knowledge that the application will never attempt to write to a read-only volume, or that any such attempt will not result in a problem, I think it's a beautiful yet risky concept to pursue at this time.

OK, I understand.

> Finally, you're never going to be able to deny access to a file/directory that's not mounted unless you resort to permissions trickery

I very much do not understand that statement... Could you please expand? Thanks

File: Dockerfile (hunk 10 above)
Comment (fabricetriboix) [reply Dockerfile]: I was just intrigued by this. That's not a problem at all.

File: Dockerfile (hunk 12 above)
Comment (fabricetriboix) [reply Dockerfile]: Done

**Review by fabricetriboix — APPROVED** [review]

File: Dockerfile (hunk 7 above)
Comment (drivera-armedia) [reply Dockerfile]: In Linux you can only deny access through permissions. So if you don't want a process to be able to access X directory for writing, the only avenue left open to you is to set the file/directory permissions such that the user running the process lacks the write access.  Ditto with read / execute.
So If you want to really sandbox a process to only have write access to a given (set of) location(s), you'll need to explicitly set the permissions appropriately. If you want to deny read access, then you have to do the same thing.
What I'm saying is that unless you invest some effort into fully documenting which areas the program _**needs**_ to be able to access (for read, write, execute, etc.), and set the rest of the permissions accordingly, you're really not going to accomplish much - i.e. a lot of effort for not enough of a return.
It's much more cost-effective to rely on volumes, coupled with containers' ephemeral nature to wipe out undesirable modifications while preserving desirable ones ...

**Review by JonHolman — APPROVED** [review]

**Review by JonHolman — APPROVED** [review]

File: Dockerfile (hunk 7 above)
Comment (fabricetriboix) [reply Dockerfile]: All right, thanks
